bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java

```diff
@@ -29,211 +34,315 @@
+     * Additional properties are "user","pwd","qos","retain","lwt","keepAlive","clientId", please read the
+     * service configuration documentation for a detailed description.
+     *
+     * @param brokerConnectionConfig The configuration key-value map.
+     * @return Returns the created broker connection.
+     * @throws ConfigurationException Most likely your provided name and url are invalid.
+     * @throws MqttException
+     */
+    public MqttBrokerConnection addBrokerConnection(Map<String, Object> brokerConnectionConfig)
+            throws ConfigurationException, MqttException {
+        String name = brokerConnectionConfig.get("name").toString();
+        if (name == null) {
+            throw new ConfigurationException("MQTT Broker property '{}' should have the format 'broker.propertykey'");
+        }
+        MqttBrokerConnection connection;
+        synchronized (brokerConnections) {
+            connection = brokerConnections.get(name.toLowerCase());
+            if (connection == null) {
+                connection = new MqttBrokerConnection(name, brokerConnectionConfig.get("url").toString());
+                brokerConnections.put(name.toLowerCase(), connection);
+                for (MqttBrokersObserver o : brokersObservers) {
+                    o.brokerAdded(connection);
+                }
+            }
+        }
+
+        connection.setCredentials(brokerConnectionConfig.get("user").toString(),
+                brokerConnectionConfig.get("pwd").toString());
+        connection.setQos((Integer) brokerConnectionConfig.get("qos"));
+        connection.setRetain((Boolean) brokerConnectionConfig.get("retain"));
+        connection.setClientId(brokerConnectionConfig.get("clientId").toString());
+        connection.setKeepAliveInterval((Integer) brokerConnectionConfig.get("keepAlive"));
+        MqttWillAndTestament will = MqttWillAndTestament.fromString(brokerConnectionConfig.get("lwt").toString());
+        if (will != null) {
+            logger.debug("Setting last will: {}", will);
+            connection.setLastWill(will);
         }
+        return connection;
     }
 
     /**
-     * Lookup an broker connection by name.
+     * Remove a broker connection
      *
-     * @param brokerName to look for.
-     * @return existing connection or new one if it didn't exist yet.
+     * @param connection The broker connection
      */
-    private MqttBrokerConnection getConnection(String brokerName) {
+    public void removeBrokerConnection(MqttBrokerConnection connection) {
         synchronized (brokerConnections) {
-            MqttBrokerConnection conn = brokerConnections.get(brokerName.toLowerCase());
-            if (conn == null) {
-                conn = new MqttBrokerConnection(brokerName);
-                brokerConnections.put(brokerName.toLowerCase(), conn);
+            brokerConnections.remove(connection.getName(), connection);
```

Comment:
This again will only work for external connections if the name is not all lower case.
In addition you may check the return value and call the observers only if the removal succeeded.